Refactor psf models

jeipollack opened the pull request:
Major Changes in this PR:

- Reorganization of most of the modules to associated packages (`script_utils.py` will be made deleted in the future) associated to issues: #44 and #46 
- Added modules for `parametric` and `semi parametric` PSF models
- Added `training_preprocessing.py` module with handlers for Training and Test data

Modules to be reviewed:

- `wf_psf/data/training_preprocessing.py`
- `wf_psf/training/train.py`

sfarrens commented on wf_psf/data/training_preprocessing.py:
```diff
@@ -0,0 +1,114 @@
+        Boolean flag to set when loading numpy files
+
+    """
+    dataset = np.load(filename, allow_pickle=allow_pickle_flag)[()]
+    return dataset
```
I think you make this one line since you don't really need the `dataset` object.

jeipollack replied:
I decided to load the file explicitly during instantiation

sfarrens commented on wf_psf/data/training_preprocessing.py:
```diff
@@ -0,0 +1,114 @@
+class TrainingDataHandler:
+    """Training Data Handler.
+
+    maybe rename to TensorFlow Handler
+    or TrainingDataSetup
```
Is this something you want to keep in the docstring? If this is not something you plan to implement in this PR then perhaps it would be better as a developer note (i.e. `# Maybe rename ...`) or a `Todo` section.

jeipollack replied:
I forgot about this note. I removed it.

jeipollack commented:
lots of commits bc I forgot I need to use my personal access token to `git push` from JZ.

sfarrens reviewed (commented):
Hi @jeipollack things are looking good. I opened a couple of minor threads. Some points to consider for `train.py`:

- The class method `train` is very long. I feel like a lot of these steps could be broken up into smaller "private" class methods.
- There are lots of `print` statements. Perhaps you could include a `verbose` option (which is `True` by default) in the UI and add `if verbose:` (or something to that effect) to these.

jeipollack commented:
Hi @sfarrens , I am ready for you to complete your review of `train.py` . The refactoring resulted in changes to `configs/training_config.yaml` and additional attributes to `psf_models/psf_models_semiparametric.py`.

jeipollack commented:
PR Msg to test GH -> Slack.

jeipollack commented on wf_psf/training/train.py:
```diff
@@ -12,11 +12,14 @@
-import wf_psf.io as io
-from wf_psf.psf_models import psf_models
+import wf_psf.utils.io as io
+from wf_psf.psf_models import *
```
changed it.

jeipollack commented on wf_psf/psf_models/psf_model_semiparametric.py:
```diff
@@ -0,0 +1,381 @@
+
+        Parameters
+        ----------
+        input_positions:
```
This function was written by @tobias-liaudat. It's not used in training but for metrics analysis.  It looks like it will accept as input a tensor in tensor flow float units. but, I will know when I test it further when refactoring the `metrics` package.

jeipollack commented on wf_psf/training/train.py:
```diff
@@ -100,6 +103,94 @@ def training_hparams(self):
+
+        Returns
+        -------
+        multi_cycle_params: type
```
fixed.

jeipollack commented on wf_psf/psf_models/psf_model_semiparametric.py:
```diff
@@ -0,0 +1,381 @@
+    ----------
+    model_params: type
+        Recursive Namespace object containing parameters for this PSF model class
+    training_params: type
```
Replaced with Recursive Namespace.

jeipollack commented on wf_psf/psf_models/psf_model_semiparametric.py:
```diff
@@ -0,0 +1,381 @@
+        beta = tf.constant(
+            np.linalg.inv(self.tf_np_poly_opd.alpha_mat) @ beta_tilde, dtype=tf.float32
+        )
+        # Get the projection for the unmixed features
+
```
added it to the To Do list to figure out what this comment is about.

jeipollack commented on wf_psf/psf_models/psf_model_semiparametric.py:
```diff
@@ -0,0 +1,381 @@
+            self.tf_np_poly_opd.alpha_mat, self.tf_np_poly_opd.S_mat, axes=1
+        )
+        # Get beta tilde as the protection of the first n_param_poly_terms (6 for d_max=2) onto the first n_zernikes.
+        beta_tilde_inner = np.array(
+            [
+                [
+                    PI_zernikes(tf_zernike_cube[j, :, :], S_tilde_slice, n_pix_zernike)
+                    for j in range(self.n_zernikes)
+                ]
+                for S_tilde_slice in S_tilde[
+                    : self.tf_poly_Z_field.coeff_mat.shape[1], :, :
+                ]
+            ]
+        )
```
added to the To Do list

jeipollack commented on wf_psf/psf_models/psf_model_semiparametric.py:
```diff
@@ -0,0 +1,381 @@
+    TF_NP_poly_OPD,
+    TF_batch_mono_PSF,
+    TF_physical_layer,
+)
```
Thanks! Done.

jeipollack commented:
hi @sfarrens I believe I've addressed all your comments and those by @tobias-liaudat on change some config variable names. Re-ran the code and all is good.
